include/oneapi/dpl/pstl/unseq_backend_simd.h

```diff
@@ -548,12 +548,16 @@ typename ::std::enable_if<is_arithmetic_plus<_Tp, _BinaryOperation>::value, ::st
 __simd_scan(_InputIterator __first, _Size __n, _OutputIterator __result, _UnaryOperation __unary_op, _Tp __init,
             _BinaryOperation, /*Inclusive*/ ::std::false_type)
 {
-    _ONEDPL_PRAGMA_SIMD_SCAN(+ : __init)
+    _Tp __cur_scan_value;
+    _ONEDPL_PRAGMA_SIMD_SCAN(+ : __cur_scan_value)
     for (_Size __i = 0; __i < __n; ++__i)
     {
-        __result[__i] = __init;
-        _ONEDPL_PRAGMA_SIMD_EXCLUSIVE_SCAN(__init)
-        __init += __unary_op(__first[__i]);
+        __cur_scan_value = __init;
+        _ONEDPL_PRAGMA_SIMD_EXCLUSIVE_SCAN(__cur_scan_value)
```

Comment:
@mmichel11, I read your description, yet I am not convinced that using `__cur_scan_value` in the scan directive is correct. Meanwhile, while this change eliminates an intra-iteration dependency, it creates a cross-iteration dependency on `__init` that is also prohibited.

Eventually, I think we need to use what practically works (even if the code formally violates the OpenMP specification), and it might be compiler-dependent. I will take a look at the CI failures, and try to make some judgement about that all. Maybe we will need to ask the compiler team experts for a recommendation.

Reply:
Yes, I see the cross-iteration dependency now. Furthermore, the [OpenMP 5.2 spec](https://www.openmp.org/spec-html/5.2/openmpse28.html#x92-960005.6) states that list items should not be modified in the scan phase which my change also violates.

I am unsure if there is an OpenMP compliant way to do this. I agree with your suggestion that we should ask the compiler team for a recommendation on this case.

Reply:
To add some of our offline discussion into this conversation to the github PR,  I had suggested reforming it to cache the element of `__first[__i]` rather than the accumulation variable, and adding the temporary as a list item.
```
    _ONEDPL_PRAGMA_SIMD_SCAN(+ : __init, __tmp)
    for (_Size __i = 0; __i < __n; ++__i)
    {
        __tmp = __first[__i];
        __result[__i] = __init;
        _ONEDPL_PRAGMA_SIMD_EXCLUSIVE_SCAN(__init, __tmp)
        {
            __init += __unary_op(__tmp);
        }
    }
```
This *almost* works, as with the extra list item `__tmp` it allows `__tmp` to have an intra-iteration dependency across the pragma, but it also disobeys the provision mentioned above that "list items should not be modified in the scan phase"

---
I don't see a way to fit this into the simd scan specification (though in practice, the implementation seems to currently work through our testing). 

Effectively 4 things need to happen in an iteration:

1. capture `__first[i]`
2. capture previous running accumulation `__init`
3. write (2) to `__result[i]`
4. increment running accumulation with (1)

The dependencies are 
[1->4], [2->3], and when in-place, we add [1->3] so we don't write over our input before capturing it.  

For omp simd exclusive scan iteration we have: 
scan phase (apply / write the accumulation)
#pragma 
input phase (do accumulation)

I believe the scan phase should include (3), and the input phase should include (4).  However, with the paraphrased provisions, 
(a) intra-iterations between scan and input phase must not exist except with list items
(b) list items should not be modified in the scan phase

This means that (1) must also be in the input phase, which would force it to be after (3), breaking the dependency added from in-place.

Reply:
I agree that in-place exclusive scan cannot be expressed in a spec compliant way. However, I believe it is vectorizable, with the algorithm being like this (for the vector of size 4):

1. capture `__first[i..i+3]` into a vector `v[0..3]`
2. build a vector `sv = {__init, __init+v[0], __init+v[0]+v[1], __init+v[0]+v[1]+v[2]}`
3. write `sv` to `__result[i..i+3]
4. `__init += v[0]+v[1]+v[2]+v[3]`
5. `i+=4`

and it sort of matches the code that Dan has suggested:

```
for( int i=0; i<N; ++i) {
    auto v = __first[i];  // input phase?
    __result[i] = __init; // scan phase
    __init += v;          // input phase
}
```
Also note that in the above vectorization algorithm and a one for inclusive scan would only differ in the step 2, by shifting values in `sv` to the "left", and appending by the same value as computed in step 4.

Reply:
Yes, I agree with your proposed algorithm that this code is vectorizable. If we are unable to have a guarantee in the spec (or an update to it), then I believe this could be something we could look to write ourselves with OpenMP simd directives. I also have messaged a contact on the compiler team to see if there are are any compiler-specific guarantees about this behavior.

In the short-term, what is the best step forward? Should we leave the known limitation with a release note so vectorized code can still be ran for the out-of-place case for exclusive_scan, or would it be best to disable this to ensure correctness in the in-place case? This was listed as a known issue for the previous release with the icpc compiler.

Reply:
> In the short-term, what is the best step forward?

It's clear that we will not fix `__simd_scan` for this release. Let's try to disable it for the in-place exclusive scan (I believe that the inclusive scan should be fine, correct me if not). But we should use an approach similar to `__iterators_possibly_equal `, instead of comparing addresses.

Reply:
In that case, I think it is best to move `__iterators_possibly_equal` into the general utils header and directly invoke it as it provides a general iterator equality case we can reuse here. 

I have made this change.

Reply:
I created an issue for implementation of a vectorized brick, #1190 which is now outside the scope of this PR. So I am resolving this thread of discussion for this PR.